skill/wallpaper.py

```diff
@@ -76,30 +96,47 @@ def next(self):
                 index_of_selected = self.collection.index(self.selected)
                 self.selected = self.collection[index_of_selected + 1]
         else:
-            self.selected = self.default
+            self.selected = self.collection[0]
 
-    def add(self, wallpaper_url: str):
-        """Adds a new wallpaper to the collection and selects it for display.
+        self.file_name_setting = self.selected.name
 
-        Args:
-            wallpaper_url: the url to download the wallpaper image from
-        """
-        response = requests.get(wallpaper_url)
-        file_path = self.skill_data_directory.joinpath("custom-wallpaper.jpg")
+
+    def _add_custom(self):
+        """Adds a new wallpaper to the collection and selects it for display."""
+        image = _download_from_url(self.url_setting)
+        file_path = self.skill_data_directory.joinpath(CUSTOM_WALLPAPER)
         with open(file_path, "wb") as wallpaper_file:
-            wallpaper_file.write(response.content)
+            wallpaper_file.write(image)
+        self._validate_custom()
         self.collect()
         self.selected = file_path
 
-    def _get_file_path(self, file_name: str):
-        """Determines if a wallpaper file name is from the user or skill directory."""
-        skill_path = self.skill_directory.joinpath(file_name)
-        skill_data_path = self.skill_data_directory.joinpath(file_name)
-        if skill_path.exists():
-            file_path = skill_path
-        elif skill_data_path.exists():
-            file_path = skill_data_path
+    def set(self):
+        """Used to set the wallpaper on boot or skill reload."""
+        file_path = self._determine_wallpaper_path()
+        if file_path.exists():
+            self.selected = file_path
+            LOG.info("Home screen wallpaper set to " + str(file_path))
+        else:
+            raise WallpaperError(
+                "no wallpaper exists matching the wallpaper current settings"
+            )
+        if file_path.name == CUSTOM_WALLPAPER:
+            self._validate_custom()
+
+    def _determine_wallpaper_path(self):
+        """Build the right file path for the wallpaper in use."""
+        if self.file_name_setting is None:
+            file_path = self.skill_directory.joinpath(DEFAULT_WALLPAPER)
+        elif self.file_name_setting == CUSTOM_WALLPAPER:
+            file_path = self.skill_data_directory.joinpath(CUSTOM_WALLPAPER)
         else:
-            file_path = None
+            file_path = self.skill_directory.joinpath(self.file_name_setting)
 
         return file_path
+
+    def _validate_custom(self):
+        """Ensure that a downloaded wallpaper is a valid image file."""
+        file_path = self.skill_data_directory.joinpath(CUSTOM_WALLPAPER)
+        if imghdr.what(file_path) != "jpeg":
```

Comment:
A good question for @AIIX - what image formats should already be supported?

Are we artificially limiting to jpeg only?

Reply:
There isn't an actual limit on the image format, All standard formats should be supported fine JPEG, PNG, SVG, GIF etc https://doc.qt.io/qt-5/qml-qtquick-image.html#details, tho GIF's do require AnimatedImage{} instead of Image{} QML component if you do want to support GIF playback for which one needs to setup their own background property.

Reply:
Should we support a GIF as wallpaper though, even if the QML supports it?  I will add PNG and SVG to the code.

Reply:
Also, SVG files are not supported by the python imghdr library https://docs.python.org/3/library/imghdr.html

Reply:
I think we should just support JPEG for this release.  The file type of custom wallpaper is hardcoded as `.jpg`.  We can add support of other image types later.